package/android/cpp/rnskia-android/SkiaOpenGLSurfaceFactory.cpp

```diff
@@ -0,0 +1,161 @@
+#include "SkiaOpenGLHelper.h"
+#include <SkiaOpenGLSurfaceFactory.h>
+
+namespace RNSkia {
+
+/**
+ * The first context created will be considered the parent / shared context and
+ * will be used as the parent / shareable context when creating subsequent
+ * contexts.
+ */
+SkiaOpenGLContext ThreadContextHolder::SharedSkiaOpenContext;
+
+/**
+ * This is the context that is static and local to each thread.
+ */
+thread_local SkiaOpenGLContext ThreadContextHolder::ThreadSkiaOpenGlContext;
+
+sk_sp<SkSurface> SkiaOpenGLSurfaceFactory::makeOffscreenSurface(int width,
+                                                                int height) {
+  RNSkLogger::logToConsole("makeOffscreenSurface Setting up Skia");
+  // Setup OpenGL and Skia:
+  if (!SkiaOpenGLHelper::createSkiaDirectContext(
+          &ThreadContextHolder::ThreadSkiaOpenGlContext,
+          &ThreadContextHolder::SharedSkiaOpenContext)) {
+    RNSkLogger::logToConsole(
+        "Could not create Skia Surface from native window / surface. "
+        "Failed creating Skia Direct Context");
+    return nullptr;
+  }
+
+  auto colorType = kN32_SkColorType;
+
+  SkSurfaceProps props(0, kUnknown_SkPixelGeometry);
+
+  // Create texture
+  auto texture =
+      ThreadContextHolder::ThreadSkiaOpenGlContext.directContext
+          ->createBackendTexture(width, height, colorType, GrMipMapped::kNo,
+                                 GrRenderable::kYes);
+
+  struct ReleaseContext {
+    SkiaOpenGLContext *context;
+    GrBackendTexture texture;
+  };
+
+  auto releaseCtx = new ReleaseContext(
+      {&ThreadContextHolder::ThreadSkiaOpenGlContext, texture});
+
+  // Create a SkSurface from the GrBackendTexture
+  return SkSurface::MakeFromBackendTexture(
+      ThreadContextHolder::ThreadSkiaOpenGlContext.directContext.get(), texture,
+      kTopLeft_GrSurfaceOrigin, 0, colorType, nullptr, &props,
+      [](void *addr) {
+        auto releaseCtx = reinterpret_cast<ReleaseContext *>(addr);
+
+        releaseCtx->context->directContext->deleteBackendTexture(
+            releaseCtx->texture);
+      },
+      releaseCtx);
+}
+
+sk_sp<SkSurface> WindowSurfaceHolder::getSurface() {
+  if (_skSurface == nullptr) {
+
+    // Do we need to setup OpenGL and Skia?
+    if (ThreadContextHolder::ThreadSkiaOpenGlContext.glDisplay ==
```

Comment:
makes sense I think. Display should probably be only available on the top level skia opengl context.